qiskit/transpiler/passes/routing/layout_transformation.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Map (with minimum effort) a DAGCircuit onto a `coupling_map` adding swap gates."""
+from typing import Union
+
+import numpy as np
+
+from qiskit.transpiler import Layout, CouplingMap
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+from qiskit.transpiler.passes.routing.algorithms import ApproximateTokenSwapper
+
+
+class LayoutTransformation(TransformationPass):
+    """ Adds a Swap circuit for a given (partial) permutation to the circuit.
+
+    This circuit is found by a 4-approximation algorithm for Token Swapping.
+    More details are available in the routing code.
+    """
+
+    def __init__(self, coupling_map: CouplingMap,
+                 from_layout: Union[Layout, str],
+                 to_layout: Union[Layout, str],
+                 seed: Union[int, np.random.RandomState] = None,
+                 trials=4):
+        """LayoutTransformation initializer.
+
+        Args:
+            coupling_map (CouplingMap):
+                Directed graph represented a coupling map.
+
+            from_layout (Union[Layout, str]):
+                The starting layout of qubits onto physical qubits.
+                If the type is str, look up `property_set` when this pass runs.
```

Comment:
I don't quite get the "str" layout.. what does that look like and why is it needed?

Reply:
I'm seeing the use cases that from/to layout is given depending on the previous pass. For example, when we want to change layout back to the initial_layout for easy verification of the mapped circuit, we'll likely to define our own pass manager like
```
my_pass_manager = [
...,
ABCSwap(coupling),
LayoutTransformation(from_layout="final_layout", to_layout="layout")
]
```